### ISSUE TITLE
Add keyboard shortcuts to fast scroll through editor tabs

### ISSUE THREAD

**Issue by madwareru** [defold/defold#4889]
**Is your feature request related to a problem? Please describe (REQUIRED):**
At the right moment there are no way to scroll through opened editor tabs with a hot keys.
It's a common feature in a wide variety of IDE. Among others is vscode and atom IDE.

**Describe the solution you'd like (REQUIRED):**
There are two new hotkeys:
`:toggle-to-next-tab`
`Shift+Meta+']'` on Mac
`Shift+Ctrl+']'` on Windows
`Shift+Ctrl+']'` on Linux

`:toggle-to-prev-tab`
`Shift+Meta+'['` on Mac
`Shift+Ctrl+'['` on Windows
`Shift+Ctrl+'['` on Linux

When there is no opened tabs then nothing happens by pressing any of hotkeys.

When the user presses a`:toggle-to-next-tab` hotkey, one of the actions occur: 
a) When the current selection in an active tab pane is not last, we simply invoke `selectNext` method on a selection model of an active tab pane.
b) When we reach the end of an active tab-pane tabs, we search for another tab pane and, if found one, setting a focus on it. Then we invoke `selectFirst` method on a selection model of currently active tab pane.

When the user presses a`:toggle-to-prev-tab` hotkey, one of the actions occur: 
a) When the current selection id in an active tab pane is higher than zero, we simply invoke `selectPrevious` method on a selection model of an active tab pane.
b) When we reach the beginning of an active tab-pane tabs, we search for another tab pane and, if found one, setting a focus on it. Then we invoke `selectLast` method on a selection model of currently active tab pane.

**Describe alternatives you've considered (REQUIRED):**
There seemed to be no other alternatives right now rather than direct clicking on a tabs with a mouse.

**Additional context (OPTIONAL):**
I made an approach myself to do this here: https://github.com/madwareru/defold/tree/lab/editor-tabs-scrolling

Most of the times it works okay, but somehow it misses some tabs sometimes.
So I'd like to discuss it with someone who may know more.
@matgis-king May be you can find where I made a mistake

Conceptually all the works is done here:
```clojure
(defn- scroll-tabs [app-view foo-table]
  (let [fmap (fn [a tr] (when a (tr a)))
        active-pane ^TabPane (g/node-value app-view :active-tab-pane)
        selection (.getSelectionModel active-pane)
        selected-id (.getSelectedIndex selection)
        tabs (.getTabs active-pane)
        tabs-size(if (nil? tabs) 0 (.size tabs))
        tabs-split ^SplitPane (g/node-value app-view :editor-tabs-split)
        other-pane (fmap tabs-split #(find-other-tab-pane % active-pane))
        other-selection(fmap other-pane #(.getSelectionModel %))
        checker-fn (:checker-fn foo-table)
        normal-switch-fn (:normal-switch-fn foo-table)
        overlap-switch-fn (:overlap-switch-fn foo-table)
       ]
    (when (and active-pane (not= 0 tabs-size))
      (if (checker-fn selected-id tabs-size)
        (normal-switch-fn selection)
        (if (nil? other-pane)
          (overlap-switch-fn selection)
          (do
            (.requestFocus other-pane)
            (overlap-switch-fn other-selection)))))))

(handler/defhandler :toggle-to-next-tab :global
  (enabled? [app-view evaluation-context]
            (not-empty (get-active-tabs app-view evaluation-context)))
  (run [app-view]
       (scroll-tabs
         app-view
         {:checker-fn (fn [id size] (not= size (+ 1 id)))
          :normal-switch-fn #(.selectNext %)
          :overlap-switch-fn #(.selectFirst %)})))

(handler/defhandler :toggle-to-prev-tab :global
  (enabled? [app-view evaluation-context]
            (not-empty (get-active-tabs app-view evaluation-context)))
  (run [app-view]
       (scroll-tabs
         app-view
         {:checker-fn (fn [id _size] (not= 0 id))
          :normal-switch-fn #(.selectPrevious %)
          :overlap-switch-fn #(.selectLast %)})))
```

**Comment by matgis-king** [comment]
It's kind of difficult to tell what could be going wrong. Perhaps you need to add some `println` calls in the various cases where you do nothing because of `nil` values?

One thing that could maybe be an issue is that the `(.requestFocus other-pane)` will not cause the focus to change immediately - it moves focus at the next "tick" of JavaFX. If it is important that the pane has focus when `overlap-switch-fn` is called, you need to enqueue the call after the next tick using:
```clj
(ui/run-later
  (overlap-switch-fn other-selection))
```

**Comment by JCash** [comment]
Just a note that the "[" and "]" aren't visible on many keyboards.
On my swedish mac keyboard, I need to press "alt+8" to get a "[", so I wonder how it plays into these shortcuts.

**Comment by madwareru** [comment]
Well I think that we need more data on how vscode works on swedish mac keyboards then. Could you try to use this feature in atom or vscode IDE?

**Comment by dapetcu21** [comment]
I did notice that occasionally a tab would get skipped by Ctrl+Tab/Ctrl+Shift+Tab with the only solution being to close it with the mouse and open it again. This has been a long standing and hard to repro bug that I never got to reporting.

**Comment by madwareru** [comment]
> I did notice that occasionally a tab would get skipped by Ctrl+Tab/Ctrl+Shift+Tab with the only solution being to close it with the mouse and open it again. This has been a long standing and hard to repro bug that I never got to reporting.

I can't find any signs of such keymappings in editor2. Was it in editor1 times may be?

**Comment by matgis-king** [comment]
Yes, there are built-in keybindings for the `TabPane` control in JavaFX. It is also quite possible the reason it does not always work is because of a bug in the JavaFX `TabPane` control. We've had quite a few issues with the built-in JavaFX controls in the past. 😞

The JavaFX project is [now on GitHub](https://github.com/openjdk/jfx) if you're interested in exploring further. However, if your feature works most of the time and we discover the underlying bug is in JavaFX, perhaps we should treat it as a separate issue.

**Comment by britzl** [comment]
@madwareru could you create a PR for this? It will make it easier to review and pick up.
